compiler/src/dotty/tools/dotc/eff/EffectOps.scala

```diff
@@ -0,0 +1,203 @@
+package dotty.tools
+package dotc
+package eff
+
+import core.*
+import Phases.*, DenotTransformers.*, SymDenotations.*
+import Contexts.*, Names.*, Flags.*, Symbols.*, Decorators.*
+import Types.*, StdNames.*, Denotations.*
+import ast.tpd, tpd.*
+import cc.*
+import CheckEffects.*
+import Annotations.*
+
+object KillType:
+  import KillOps.*
+
+  def apply(tp: Type, refs: List[Tree])(using Context): Type =
+    val annotTree =
+      New(getKillAnnot.typeRef,
+        Typed(
+          SeqLiteral(refs, TypeTree(defn.AnyType)),
+          TypeTree(defn.RepeatedParamClass.typeRef.appliedTo(defn.AnyType))) :: Nil)
+    AnnotatedType(tp, Annotation(annotTree))
+
+  def unapply(tp: Type)(using Context): Option[(Type, List[Tree])] =
+    tp match
+      case AnnotatedType(parent, annot) if annot.symbol.isKill =>
+        Some(parent, annot.tree.killedElems)
+      case _ => None
+end KillType
+
+object KillOps:
+  /**
+   * Given a annotation, finds its killedElems from its tree (annot.tree).
+   */
+  extension (tree: Tree)
+    def killedElems(using Context): List[Tree] = tree match
+      case Apply(_, Typed(SeqLiteral(elems, _), _) :: Nil) =>
+        elems
```

Comment:
This means all actual arguments involved in application are considered killed? I guess I need to look at the usages of this method.